core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -3144,6 +3190,105 @@ private RelNode rewriteAggregateWithGroupId(Blackboard bb,
+      int arg = originArgs.get(ordinal);
+      if (newGroupList.contains(arg)) {
+        newArgs.add(arg);
+      } else {
+        ordinals.add(ordinal);
+      }
+    }
+    if (newArgs.isEmpty()) {
+      long groupingValue = 1L << originGroupSet.asList().size() - 1;
+      groupingValues.put(aggCallOrdinal, groupingValue);
+      return null;
+    }
+    final ImmutableBitSet argsDropped = ImmutableBitSet.of(ordinals);
+    groupingArgOrdinals.put(aggCallOrdinal, argsDropped);
+    final int originFilterArg = aggregateCall.filterArg;
+    final int filterArg = newGroupList.contains(originFilterArg) ? originFilterArg : -1;
+    return AggregateCall.create(aggregateCall.getAggregation(), aggregateCall.isDistinct(),
+        aggregateCall.isApproximate(), aggregateCall.ignoreNulls(), newArgs,
+        filterArg, aggregateCall.collation, aggregateCall.type, aggregateCall.name);
+  }
+
+  /**
+   * Re-compute the initial value from grouping function (with partial arguments).
+   *
+   * <p>For example, assume the initial function is: {@code grouping(a, b, c)},
+   * when the second argument is always null (i.e., {@code b = null}), we can
+   * get the result of {@code grouping(a, b, c)} from {@code grouping(a, c)}.
+   * That is, {@code value = grouping(a, c), argOrdinals = {1}, valueLen = 2}.
+   * </p>
+   *
+   * @param groupingValue the input grouping value
+   * @param argOrdinals ordinals of grouping args dropped
+   * @param valueLen length of the input number in binary
+   * @return result of the initial grouping value
+   */
+  private RexNode reComputeGrouping(final RexNode groupingValue,
+      final ImmutableBitSet argOrdinals, final int valueLen) {
+    int resultLen = valueLen;
+    RexNode result = groupingValue;
+    for (int ordinal: argOrdinals) {
+      result = insertBitInSpecificPosition(result, ordinal, resultLen);
+      resultLen++;
+    }
+    return result;
+  }
+
+  /**
+   * Insert a bit '1' into the value at specific position
+   * For example, when input = 7 (i.e., '111'), ordinal = 2, len = 3,
+   * the result is 15 (i.e., '1111')
+   * @param input the input number
+   * @param ordinal position
+   * @param len length of the input number in binary
+   * @return number after inserting bit
+   */
+  private RexNode insertBitInSpecificPosition(final RexNode input,
+       final int ordinal, final int len) {
+    assert ordinal >= 0 && ordinal <= len;
+    final int rightLen = len - ordinal;
+    final long v = (long) Math.pow(2, rightLen);
+    final RexNode vLiteral = rexBuilder.makeBigintLiteral(BigDecimal.valueOf(v));
+    // left part: num / v
+    final RexNode left = rexBuilder.makeCall(SqlStdOperatorTable.DIVIDE, input, vLiteral);
+    // left value: left * v
+    final RexNode leftValue =
+        rexBuilder.makeCall(SqlStdOperatorTable.MULTIPLY, left, vLiteral);
+    // right value: num - leftValue
+    final RexNode right =
+        rexBuilder.makeCall(SqlStdOperatorTable.MINUS, input, leftValue);
```

Comment:
This looks too complicated. AFAIK this should be executed for literals only. Why is the math executed over RexNodes?
Is'nt regular Java's int/long/BigInteger enough?

Reply:
Because the input RexNode is `grouping` function's result, which is not a literal.
E.g., we want to calculate the initial `grouping(a, null, c)` based on `grouping(a, c)`
```
grouping(a, c)         grouping(a, null, c)
'00'                         '010'
'01'                         '011'
'10'                         '110'
'11'                         '111'
```

Reply:
Can we create a new `grouping` invocation instead?
Why is it important to do math computations?

Reply:
> Can we create a new grouping invocation instead?

We can't. Because each subquery's `groupSet` is just a sub-set of the original `groupSet`.

> Why is it important to do math computations?

The problem can we simplified as: How can we get the value of `g1` based on `g2`.

**Q1:**
```
select grouping(a, b, c) as g1
from t
group by grouping sets (a, b, c)
```
**Q2:**
```
select grouping(a, c) as g2
from t
group by grouping sets(a, c)
```

Reply:
Grouing functions are special, as they take `groupSet` fields as its input.

Reply:
Oh, I think I understand now, however, it is sad the code comments do not provide examples.

Is `grouping(a, null, b)` valid? Should we use it instead?

If `null` is not valid, then how about `g1 = (grouping(a) << 2) + grouping(c)`? I guess it would be much easier to reason about than nested multiply-divide ops.

Reply:
> Is grouping(a, null, c) valid? Should we use it instead?

`grouping(a, null, c)` is not valid, calcite runtime use arg ordinal to conduct bit operation.

> If null is not valid, then how about g1 = (grouping(a) << 2) + grouping(c)? I guess it would be much easier to reason about than nested multiply-divide ops.

1) It is `g1 = 4*grouping(a) + 2 + grouping(c)`, we mark the position as '1' when it is null.
2) Good idea! Seperating to atomic functions with single argument is intuitive. But maybe we can't avoid multiply ops, since bit shift is not supported now. A more general example.
```
grouping(a, null, null, d, e, null, g, null) =
128* grouping(a)* + 64 + 32 + 16*grouping(d) + 8*grouping(e) + 4 + 2*grouping(g) + 1
```

Reply:
`d, e` could come together in `grouping(d, e)`:

```
grouping(a, null, null, d, e, null, g, null) =
128*grouping(a) + 64 + 32 + 8*grouping(d, e) + 4 + 2*grouping(g) + 1
```

where 64+32+4+1 could be constant-folded.

Reply:
That's nice! From the Julian's comment in Jira, I also come up with another approach.
We can compensate the removed groupset members and filter them again.
```
select grouping(a, d, e, g) as g
from t
group by grouping sets(a, d, e, g)
```
rewrites to
```
select g2
from (
  select (a, b, c, d, e, f, g, h) as g, grouping(b) as g1, grouping(c) as g1, grouping(f) as g3, grouping(h) as g4
  from t
  group by grouping sets(a, b, c, d, e, f, g, h)
)
where g1 = 1 and g2 = 1 and g3 = 1 and g4 = 1
```